src/dataflow/src/source/kafka.rs

```diff
@@ -603,19 +612,51 @@ fn create_kafka_config(
     kafka_config
 }
 
-impl<'a> From<&BorrowedMessage<'a>> for SourceMessage<Option<Vec<u8>>, Option<Vec<u8>>> {
-    fn from(msg: &BorrowedMessage<'a>) -> Self {
-        let kafka_offset = KafkaOffset {
-            offset: msg.offset(),
-        };
-        Self {
-            partition: PartitionId::Kafka(msg.partition()),
-            offset: kafka_offset.into(),
-            upstream_time_millis: msg.timestamp().to_millis(),
-            key: msg.key().map(|k| k.to_vec()),
-            value: msg.payload().map(|p| p.to_vec()),
+fn construct_headers(
+    headers: Option<&BorrowedHeaders>,
+) -> Result<Vec<(String, MessagePayload)>, anyhow::Error> {
+    let mut out = Vec::new();
+    use rdkafka::message::Headers;
+
+    if let Some(headers) = headers {
+        for idx in 0..headers.count() {
+            let (k, v) = headers.get(idx).ok_or_else(|| {
+                anyhow::anyhow!(
+                    "rdkafka told us there were {} headers \
+                    but failed to provide us the {}-index header",
+                    headers.count(),
+                    idx
+                )
+            })?;
```

Comment:
I'm worried there might be a bug in rust-rdkafka where it doesn't properly handle headers with null values, because the possibility isn't represented in the type signature here.